### PR TITLE
More efficiently parse shas for shaPostProcessor

### LINKED ISSUE #16092: Postprocess performance profile
Postprocess is very slow when render a 900 line csv file. This is the cpu profile graph.

图片

### PR DESCRIPTION
**Pull request by zeripath**
The shaPostProcessor currently repeatedly calls git rev-parse --verify on both backends
which is fine if there is only one thing that matches a sha - however if there are
multiple things then this becomes wildly inefficient.

This PR provides functions for both backends which are much faster to use.

Fix #16092

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: modules/git/repo_branch_gogit.go [hunk 1]
```diff
@@ -13,6 +13,30 @@ import (
 	"github.com/go-git/go-git/v5/plumbing"
 )
 
+// IsReferenceExist returns true if given reference exists in the repository.
```
Comment (zeripath) [inline modules/git/repo_branch_gogit.go]:
```suggestion
// IsObjectExist returns true if given reference exists in the repository.
```

File: modules/git/repo_branch_nogogit.go [hunk 2]
```diff
@@ -9,10 +9,28 @@ package git
 
 import (
 	"bufio"
+	"bytes"
 	"io"
 	"strings"
 )
 
+// IsReferenceExist returns true if given reference exists in the repository.
```
Comment (zeripath) [inline modules/git/repo_branch_nogogit.go]:
```suggestion
// IsObjectExist returns true if given reference exists in the repository.
```

**Comment by zeripath** [comment]
Sometimes I really hate our linters

File: modules/markup/renderer.go [hunk 3]
```diff
@@ -42,6 +43,32 @@ type RenderContext struct {
...
+		return
+	}
+	oldCancelFn := ctx.cancelFn
+	if oldCancelFn == nil {
+		ctx.cancelFn = fn
+		return
+	}
+	ctx.cancelFn = func() {
+		defer oldCancelFn()
+		fn()
```
Comment (lunny) [inline modules/markup/renderer.go:75]: Just use 
```
fn()
oldCancelFn()
```

File: modules/markup/html.go [hunk 4]
```diff
@@ -947,6 +948,17 @@ func sha1CurrentPatternProcessor(ctx *RenderContext, node *html.Node) {
...
 	hash := node.Data[m[2]:m[3]]
+
+	if ctx.GitRepo == nil {
+		var err error
+		ctx.GitRepo, err = git.OpenRepository(ctx.Metas["repoPath"])
+		if err != nil {
+			log.Error("unable to open repository: %s Error: %v", ctx.Metas["repoPath"], err)
+			return
+		}
+		ctx.AddCancel(ctx.GitRepo.Close)
```
Comment (lunny) [inline modules/markup/html.go]: And also set `ctx.GitRepo = nil`?

**Comment by lunny** [comment]
And please resolve the conflicts

File: modules/markup/renderer.go (hunk 3 above)
Comment (zeripath) [reply modules/markup/renderer.go:75]: what if fn() panics?

**Comment by zeripath** [comment]
conflicts resolved

**Comment by zeripath** [comment]
conflicts resolved again

**Review by 6543 — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Pull request merged**